Frameworks/CoreGraphics/CGGradient.mm

```diff
@@ -18,186 +18,170 @@
     }
 
-    _count = count;
+    SetLocations(locations, count);
 }
 
 /**
  @Status Interoperable
 */
-CGGradientRef CGGradientCreateWithColorComponents(CGColorSpaceRef colorSpace,
-                                                  const float* components,
-                                                  const float* locations,
-                                                  size_t count) {
-    CGGradientRef ret = new __CGGradient();
-    ret->initWithColorComponents(components, locations, count, colorSpace);
-
-    return ret;
+CFTypeID CGGradientGetTypeID() {
+    return __CGGradient::GetTypeID();
 }
 
 /**
  @Status Interoperable
 */
-CGGradientRef CGGradientCreateWithColors(CGColorSpaceRef colorSpace, CFArrayRef colors, const float* locations) {
-    CGGradientRef ret = new __CGGradient();
-    ret->initWithColors(colors, locations, colorSpace);
-
-    return ret;
+void CGGradientRelease(CGGradientRef gradient) {
+    RETURN_IF(!gradient);
+    CFRelease(static_cast<CFTypeRef>(gradient));
 }
 
 /**
  @Status Interoperable
 */
-void CGGradientRelease(CGGradientRef gradient) {
-    CFRelease((id)gradient);
+CGGradientRef CGGradientRetain(CGGradientRef gradient) {
+    RETURN_NULL_IF(!gradient);
+    CFRetain(static_cast<CFTypeRef>(gradient));
+    return gradient;
+}
+
+static inline bool __isValidColorspaceModel(CGColorSpaceRef colorspace) {
+    RETURN_RESULT_IF_NULL(colorspace, false);
+
+    // The colorspace cannot be kCGColorSpaceModelPattern || kCGColorSpaceModelIndexed
+    CGColorSpaceModel colorSpaceModel = CGColorSpaceGetModel(colorspace);
+    if ((colorSpaceModel == kCGColorSpaceModelPattern) || (colorSpaceModel == kCGColorSpaceModelIndexed)) {
+        return false;
+    }
+
+    return true;
 }
 
 /**
  @Status Interoperable
 */
-CGGradientRef CGGradientRetain(CGGradientRef gradient) {
-    CFRetain((id)gradient);
+CGGradientRef CGGradientCreateWithColorComponents(CGColorSpaceRef colorSpace,
+                                                  const CGFloat* components,
+                                                  const CGFloat* locations,
+                                                  size_t count) {
+    RETURN_NULL_IF(!colorSpace);
+    RETURN_NULL_IF(!components);
+    RETURN_NULL_IF(count == 0);
+    RETURN_NULL_IF(!__isValidColorspaceModel(colorSpace));
+    // location can be null
+    CGGradientRef gradient = new __CGGradient(colorSpace);
+    gradient->InitWithColorComponents(components, locations, count);
     return gradient;
 }
 
 /**
- @Status Stub
- @Notes
+ @Status Interoperable
 */
-CFTypeID CGGradientGetTypeID() {
-    UNIMPLEMENTED();
-    return StubReturn();
+CGGradientRef CGGradientCreateWithColors(CGColorSpaceRef colorSpace, CFArrayRef colors, const CGFloat* locations) {
+    RETURN_NULL_IF(!colorSpace);
+    RETURN_NULL_IF(!colors);
+    RETURN_NULL_IF(!__isValidColorspaceModel(colorSpace));
+    // location can be null
+    CGGradientRef gradient = new __CGGradient(colorSpace);
```

Comment:
I debated in my head.

---
In reply to: [106462392](https://github.com/Microsoft/WinObjC/pull/2249#discussion_r106462392) [](ancestors = 106462392)